### PR TITLE
[ntuple] Implement foundations for late model extension (1/2)

### PR DESCRIPTION
**Pull request by jalopezg-git**
This pull request implements the foundations for late model extension, i.e. extending the ntuple model with new fields after the initial schema has been fixed.
Per this PR, we only allow extending the model before the first entry is filled.  This restriction will be removed in a follow-up PR.

**NOTE:** after the changes in this PR, all the currently existing ntuple files remain readable.

The user-facing API is based on a new class `RNTupleModel::RUpdater` that can be used as follows:
```c++
auto model = RNTupleModel::Create();
auto field1 = model->MakeField<float>("field1", 42.0);

auto ntuple = RNTupleWriter::Recreate(std::move(model), "myNTuple", "some/file.root");
auto modelUpdater = ntuple->CreateModelUpdater();
modelUpdater->BeginUpdate();
float field2 = 10.0;
modelUpdater->AddField<float>("field2", &field2);
modelUpdater->CommitUpdate();
// possibly more calls to `BeginUpdate()`/`CommitUpdate()` here...

*field2 = 42.0;
ntuple->Fill();
```

## Changes or fixes:
- Changes required to `RNTupleDescriptor`: in particular, _(i)_ `RNTupleDescriptorBuilder::BeginHeaderExtension()` marks the beginning of the header extension; and _(ii)_ extension header information, including an iterator over the top-level fields may be obtained via `RNTupleDescriptor::GetHeaderExtension()`.
- Serialize/deserialize an incremental schema description, i.e. schema extension, as part of the footer envelope.  This changes the layout of the footer as explained in the [updated binary format specification](https://github.com/jalopezg-git/root/blob/2ed0dd65698c937f0c06017172c6fbcdfbbb9636/tree/ntuple/v7/doc/specifications.md).
**NOTE:** another change required in the serialization code is the use depth-first serialization of columns instead.  Rationale: given that
  1. New columns may appear during write as a result of schema extension; and
  2. On-disk IDs for those columns have to be assigned before serializing the next page list,

  we have to ensure that these columns get the same ID both when they are seen for the first time (in page lists) and by the time the description for the schema extension is serialized.
  This change only affects how the schema is serialized and does not change deserialization, i.e. the change is backwards compatible.
- `RPageSink::UpdateSchema(const RNTupleModelChangeset &)` allows incremental updates to the ntuple descriptor kept internally in the `RPageSink`, which is required, e.g. if new fields were added after the initial call to `RPageSink::Create(RNTupleModel &)`.
- Per the current state of affairs in `RPageSinkBuf`, some adjustments are also needed to keep `fInnerModel` in sync.
- Add class `RNTupleModel::RUpdater`, which provides limited support for incremental updates, e.g. addition of new fields.
- Add `RNTupleWriter::CreateModelUpdater()` that returns a usable `RUpdater` object.

## Checklist:
- [X] tested changes locally
- [X] updated the docs (if necessary)
- [ ] Update to format version to RC 2 -- to be done after merging both parts + split encoding

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:399]: Add class documentation

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:523]: Let's remove

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:514]: Let's remove

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:399]: Since we have a public `GetHeaderExtension()`, shouldn't the class definition be public as well?

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:423]: Let's simply return a const reference to `fFields`

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx:396]: Can be removed

File: tree/ntuple/v7/src/RPageSinkBuf.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RPageSinkBuf.cxx:52]: Let's add a TODO. We should be able, in general, to simplify the buffered sink. But this is for another PR.

**Review by jblomer — CHANGES_REQUESTED** [review]
Outstanding piece of work!

Some comments inline. General suggestions:
  - Let's make the `IncrementalUpdater` a sub class (and friend) of `RNTupleModel`, and rename it to `Updater`
  - Let's also make the `RNTupleModelChangeset` a sub class of `RNTupleModel`
  - Let's split the schema serialization from the field and column id mapping. In this way, we don't need to distinguish between initial fields/columns and late fields/columns.

And yes, together with the delta+split encoding for index columns, we should bump the release candidate number to 2.

**Comment by jblomer** [comment]
> > Let's also make the `RNTupleModelChangeset` a sub class of `RNTupleModel`
> 
> This was actually intentional, as the class is forward-declared in `RPageStorage.hxx`. C++ does not allow forward-declaring a nested class.

I see, let's leave `RNTupleModelChangeset` as is then.

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx
Comment (jalopezg-git) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:150]: I think that it may be actually useful to have these two `AddField()` overloads as-is, as they provide a means to also capture a value -- avoiding an additional call to `CaptureValueUnsafe()`.

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx
Comment (jalopezg-git) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:315]: Currently, `RPageSinkBuf` calls it directly to unfreeze the inner model.  I would suggest leaving it public until after the refactor of `RPageSinkBuf`.

File: tree/ntuple/v7/doc/specifications.md
Comment (jalopezg-git) [reply tree/ntuple/v7/doc/specifications.md:404]: Here it is, indeed, "schema".

Rationale: as per the description [here](https://github.com/root-project/root/blob/8fd3f16c84751f5c56946940f4d9d6ebd8c23ba1/tree/ntuple/v7/doc/specifications.md?plain=1#L230), _schema description_ is the term that we use in the binary specification to refer list of all fields, columns, alias columns, and extra type information contained in the header.
The footer may optionally contain an extension to the schema description provided in the header, that is interpreted according to the conditions stated [here](https://github.com/root-project/root/blob/8fd3f16c84751f5c56946940f4d9d6ebd8c23ba1/tree/ntuple/v7/doc/specifications.md?plain=1#L412-L415).

In particular, this mechanism allows for unrestricted extension to the schema found in the header, i.e. the information contained there is interpreted as if it was found at the end of the principal header.
A different story though, is that, from the users' perspective, we only allow extending the RNTupleModel.  :slightly_smiling_face:

**Comment by jalopezg-git** [comment]
> * [2023-04-13T13:06:09.466Z] /data/sftnight/workspace/root-pullrequests-build/build/include/ROOT/RNTupleModel.hxx:156:10: error: no matching function for call to ‘ROOT::Experimental::RNTupleModel::AddField<std::array<double, 2> >(std::pair<std::basic_string_view<char>, std::basic_string_view<char> >&, std::array<double, 2>*&)’

This error is due to a prior merge of #12519.  @jblomer, we need to rebase; I'll do that after the second review, as I think rebasing will make the incremental review harder.

File: tree/ntuple/v7/src/RNTupleSerialize.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RNTupleSerialize.cxx:1077]: Add a comment why the descriptor's field list is never empty (the zero field is always added before `DeserializeSchemaDescription` is called)

File: tree/ntuple/v7/src/RNTupleSerialize.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RNTupleSerialize.cxx:70]: Add documentation to this parameter, whose meaning at least to me was not obvious when reading the code (it became clear when seeing how the function is used).

Can we not keep this information in the context, btw? Like: serialized up to this and that id?

File: tree/ntuple/v7/test/ntuple_descriptor.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_descriptor.cxx:149]: Let's add a projected field / alias column in the extension header, too

File: tree/ntuple/v7/test/ntuple_serialize.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_serialize.cxx:627]: Also here, let's test the projected field / alias column in the extension header, too.

File: tree/ntuple/v7/test/ntuple_schemaext.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_schemaext.cxx:169]: Ok for now. In the final version, let's find a real-world test that mimics the use in experiment frameworks.

**Review by jblomer — CHANGES_REQUESTED** [review]
Excellent! Just some minor comments.

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx
Comment (jalopezg-git) [reply tree/ntuple/v7/src/RNTupleDescriptor.cxx:325]: That's right; thanks for pointing out!

File: tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx
Comment (jalopezg-git) [reply tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx]: > ```
>    void AddPhysicalColumn();
>    void AddAliasColumn();
> ```

I have combined these two in `void AddColumn(bool isAliasColumn)`, which I find slightly more convenient.

File: tree/ntuple/v7/src/RNTupleSerialize.cxx
Comment (jalopezg-git) [reply tree/ntuple/v7/src/RNTupleSerialize.cxx:75]: > Can we not keep this information in the context, btw? Like: serialized up to this and that id?

We could, but given that these functions may be called more than once, e.g. to compute the buffer size prior to serialization, I would prefer to leave as-is.  Also, this function is considered internal.

I have documented the function though.

**Review by jblomer — APPROVED** [review]

File: tree/ntuple/v7/test/ntuple_schemaext.cxx
Comment (pcanal) [inline tree/ntuple/v7/test/ntuple_schemaext.cxx:3]: The test (and file) might be misnamed.

Comment (jalopezg-git) [reply tree/ntuple/v7/test/ntuple_schemaext.cxx:3]: Thanks for noticing :+1: .  I have added a commit to https://github.com/root-project/root/pull/12676 that renames the test.

Comment (pcanal) [reply tree/ntuple/v7/test/ntuple_schemaext.cxx:3]: Thanks.

**Pull request merged**
